Mathlib/NumberTheory/JacobiSum/Basic.lean

```diff
@@ -107,4 +106,62 @@ theorem jacobiSum_trivial_trivial :
 theorem jacobiSum_one_one : jacobiSum (1 : MulChar F R) 1 = Fintype.card F - 2 :=
   jacobiSum_trivial_trivial
 
+
+variable [IsDomain R] -- needed for `MulChar.sum_eq_zero_of_ne_one`
+
+/-- If `χ` is a nontrivial multiplicative character on a finite field `F`, then `J(1,χ) = -1`. -/
+theorem jacobiSum_one_nontrivial {χ : MulChar F R} (hχ : χ ≠ 1) : jacobiSum 1 χ = -1 := by
+  rw [jacobiSum_eq_aux, MulChar.sum_eq_zero_of_ne_one hχ, MulChar.sum_one_eq_card_units,
+    Fintype.card_eq_card_units_add_one (α := F), add_zero, Nat.cast_add, Nat.cast_one,
+    ← sub_sub, sub_self, zero_sub, add_right_eq_self]
+  convert sum_const_zero with x hx
+  simp only [mem_sdiff, mem_univ, mem_insert, mem_singleton, not_or, true_and] at hx
+  rw [MulChar.one_apply <| isUnit_iff_ne_zero.mpr hx.1, sub_self, zero_mul]
+
+/-- If `χ` is a nontrivial multiplicative character on a finite field `F`,
+then `J(χ,χ⁻¹) = -χ(-1)`. -/
+theorem jacobiSum_nontrivial_inv {χ : MulChar F R} (hχ : χ ≠ 1) : jacobiSum χ χ⁻¹ = -χ (-1) := by
+  rw [jacobiSum]
+  conv => enter [1, 2, x]; rw [MulChar.inv_apply', ← map_mul, ← div_eq_mul_inv]
+  rw [sum_eq_sum_diff_singleton_add (mem_univ (1 : F)), sub_self, div_zero, χ.map_zero, add_zero]
+  have : ∑ x ∈ univ \ {1}, χ (x / (1 - x)) = ∑ x ∈ univ \ {-1}, χ x := by
+    refine sum_bij' (fun a _ ↦ a / (1 - a)) (fun b _ ↦ b / (1 + b)) (fun x hx ↦ ?_)
+      (fun y hy ↦ ?_) (fun x hx ↦ ?_) (fun y hy ↦ ?_) (fun _ _ ↦ rfl)
+    · simp only [mem_sdiff, mem_univ, mem_singleton, true_and] at hx ⊢
+      rw [div_eq_iff <| sub_ne_zero.mpr ((ne_eq ..).symm ▸ hx).symm, mul_sub, mul_one,
+        neg_one_mul, sub_neg_eq_add, self_eq_add_left, neg_eq_zero]
+      exact one_ne_zero
+    · simp only [mem_sdiff, mem_univ, mem_singleton, true_and] at hy ⊢
+      rw [div_eq_iff fun h ↦ hy <| eq_neg_of_add_eq_zero_right h, one_mul, self_eq_add_left]
+      exact one_ne_zero
+    · simp only [mem_sdiff, mem_univ, mem_singleton, true_and] at hx
+      rw [eq_comm, ← sub_eq_zero] at hx
+      field_simp
+    · simp only [mem_sdiff, mem_univ, mem_singleton, true_and] at hy
+      rw [eq_comm, neg_eq_iff_eq_neg, ← sub_eq_zero, sub_neg_eq_add] at hy
+      field_simp
+  rw [this, ← add_eq_zero_iff_eq_neg, ← sum_eq_sum_diff_singleton_add (mem_univ (-1 : F))]
+  exact MulChar.sum_eq_zero_of_ne_one hχ
```

Comment:
Not sure if this is an improvement, but I factored out that bijection:
```suggestion
lemma Set.BijOn.of_inverse {α β : Type*} {s : Set α} {t : Set β} (f : α → β) (g : β → α)
    (hf : MapsTo f s t) (hg : MapsTo g t s)
    (hfg : ∀ x ∈ s, g (f x) = x) (hgf : ∀ y ∈ t, f (g y) = y) :
    Set.BijOn f s t := by
  refine ⟨hf, ?_, ?_⟩
  · intro x hx y hy h
    rw [← hfg x hx, ← hfg y hy, h]
  · intro x hx
    use g x
    exact ⟨hg hx, hgf _ hx⟩

lemma div_sub_bijOn (a : F) (ha : a ≠ 0) :
    Set.BijOn (fun x ↦ x / (a - x)) (univ \ {a} : Finset F) (univ \ {-1} : Finset F) := by
  apply Set.BijOn.of_inverse _ (fun y ↦ a * y / (1 + y))
  all_goals
    intro x hx
    simp only [mem_coe, mem_sdiff, mem_univ, mem_singleton, true_and] at hx ⊢
  · rw [eq_comm, ← sub_eq_zero] at hx
    field_simp
    rw [← sub_eq_zero]
    ring_nf
    exact ha
  · rw [eq_neg_iff_add_eq_zero, add_comm] at hx
    field_simp [ha]
  · rw [eq_comm, ← sub_eq_zero] at hx
    field_simp
  · rw [eq_neg_iff_add_eq_zero, add_comm] at hx
    field_simp
    ring_nf
    field_simp

lemma Finset.sum_bijOn {α β A : Type*} [AddCommMonoid A]
    {s : Finset α} {t : Finset β} (f : α → β) (g : β → A) (h : Set.BijOn f s t) :
    ∑ x in s, g (f x) = ∑ y in t, g y :=
  sum_nbij f h.mapsTo h.injOn h.surjOn (fun _ _ ↦ rfl)

/-- If `χ` is a nontrivial multiplicative character on a finite field `F`,
then `J(χ,χ⁻¹) = -χ(-1)`. -/
theorem jacobiSum_nontrivial_inv {χ : MulChar F R} (hχ : χ ≠ 1) : jacobiSum χ χ⁻¹ = -χ (-1) := by
  rw [jacobiSum]
  conv => enter [1, 2, x]; rw [MulChar.inv_apply', ← map_mul, ← div_eq_mul_inv]
  have : ∑ x ∈ univ \ {1}, χ (x / (1 - x)) = ∑ x ∈ univ \ {-1}, χ x :=
    Finset.sum_bijOn _ _ (div_sub_bijOn 1 one_ne_zero)
  rw [sum_eq_sum_diff_singleton_add (mem_univ (1 : F)), this, sum_sdiff_eq_sub (subset_univ _),
    MulChar.sum_eq_zero_of_ne_one hχ]
  simp [χ.map_zero]
```
I really wish that automation would be better at proving the bijection.

Reply:
I think the addition of the helper lemmas is better done in a separate PR; after that is merged, we can refactor.